### PR TITLE
Uncomment code that is supposedly working now due to Rust 1.51.

### PR DESCRIPTION
**Pull request by CGMossa**
I found this comment in a `filter.rs` file, and I'd like to see if this works and what it does. 

This is just a draft. 
I had to make the PR to invoke the Actions.

### CODE REVIEW

File: crates/bevy_ecs/src/query/filter.rs [hunk 1]
```diff
@@ -12,14 +12,13 @@ use std::{marker::PhantomData, ptr};
 
 // TODO: uncomment this and use as shorthand (remove where F::Fetch: FilterFetch everywhere) when
```
Comment (bjorn3) [inline crates/bevy_ecs/src/query/filter.rs:13]: You also need to remove the `F::Fetch: FilterFetch` everywhere.

Comment (mockersf) [reply crates/bevy_ecs/src/query/filter.rs:13]: here for example: https://github.com/bevyengine/bevy/blob/b07db8462f20fff9f5f6c1af915ad73700b7a264/crates/bevy_ecs/src/query/iter.rs#L14

**Comment by bjorn3** [comment]
You will need to run `cargo fmt`.

**Comment by CGMossa** [comment]
My development environment is shattered. I'm on Windows, am I supposed to have this configuration in git?
```
git config --global core.autocrlf input
```

Because unless I have this on, `cargo fmt` was chaging every single file.

**Comment by CGMossa** [comment]
I'm really sorry, but could someone explain this to me? 

```rust
type mismatch resolving `for<'a> <<F as query::fetch::WorldQuery>::Fetch as query::fetch::Fetch<'a>>::Item == bool`
        expected type `bool`
found associated type `<<F as query::fetch::WorldQuery>::Fetch as query::fetch::Fetch<'_>>::Item`
consider constraining the associated type `<<F as query::fetch::WorldQuery>::Fetch as query::fetch::Fetch<'_>>::Item` to `bool`
for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
```

**Comment by cart** [comment]
> Because unless I have this on, cargo fmt was changing every single file.

cargo fmt should always use "unix" line endings as we have explicitly configured it to that. Maybe the default windows git settings caused git to immediately overwrite the "unix" line endings with CLRF?

**Comment by CGMossa** [comment]
> > Because unless I have this on, cargo fmt was changing every single file.
> 
> cargo fmt should always use "unix" line endings as we have explicitly configured it to do that. Maybe the default windows git settings caused git to immediately overwrite the "unix" line endings with CLRF?

I think that's what was happening. In anyways, I've changed this, so it should be fine. 
I honestly don't get how to solve the current error I'm getting. Thus far, I think I need to constrain
an `Item` in `Fetch` to be bool due to something, but I don't know how to do that in a meaningful way.

**Comment by mockersf** [comment]
> but I don't know how to do that in a meaningful way

I took a quick look, didn't find an easy fix. Maybe @cart was over-optimistic in that it would work?

**Comment by cart** [comment]
Hmm honestly I can't tell if this is a rustc bug, a fundamental rust trait limitation, or a "trait declaration bug" on our part. Here is a minimal reproduction of the issue:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c916a457ff042216c64d2f36cc893cc9

I don't understand why the "bool" constraint even comes into the equation when declaring `Query`. There could be any number of impls for FilterFetch. Is the "bool" message just a case of "rustc trying to provide a helpful error message by looking at potential impls"?

**Comment by bjorn3** [comment]
When using `-Zverbose` I get the following error:

```rust
error[E0271]: type mismatch resolving `<<F as WorldQuery>::Fetch as Fetch>::Item == bool`
  --> <anon>:23:44
   |
9  | trait FilterQuery: WorldQuery where Self::Fetch: FilterFetch {}
   |                                                  ----------- required by this bound in `FilterQuery`
...
23 | struct Query<Q, F> where Q: WorldQuery, F: FilterQuery {
   |                                            ^^^^^^^^^^^ expected `bool`, found associated type
   |
   = note:         expected type `bool`
           found associated type `<<F as WorldQuery>::Fetch as Fetch>::Item`
   = help: consider constraining the associated type `<<F as WorldQuery>::Fetch as Fetch>::Item` to `bool`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
note: required because of the requirements on the impl of `FilterFetch` for `<F as WorldQuery>::Fetch`
  --> <anon>:15:29
   |
15 | impl<T: Fetch<Item = bool>> FilterFetch for T {
```

**Comment by Guvante** [comment]
It looks like [implied bounds](https://github.com/rust-lang/rfcs/blob/master/text/2089-implied-bounds.md) is the name of the RFC that this needs.

And the [issue](https://github.com/rust-lang/rust/issues/44491) points out that chalk already supports implied bounds so there are no plans on backporting the feature to rustc. Instead it is blocked on chalk integration.

If you remove the blanket impl (`impl<T: Fetch<Item = bool>> FilterFetch for T`) in your playground link you will get the following error which is the real error that is being raised.

```
error[E0277]: the trait bound `<F as WorldQuery>::Fetch: FilterFetch` is not satisfied
  --> src/lib.rs:38:44
   |
9  | trait FilterQuery: WorldQuery where Self::Fetch: FilterFetch {
   |                                                  ----------- required by this bound in `FilterQuery`
...
38 | struct Query<Q, F> where Q: WorldQuery, F: FilterQuery {
   |                                            ^^^^^^^^^^^ the trait `FilterFetch` is not implemented for `<F as WorldQuery>::Fetch`
   |
help: consider extending the `where` bound, but there might be an alternative better way to express this requirement
   |
38 | struct Query<Q, F> where Q: WorldQuery, F: FilterQuery, <F as WorldQuery>::Fetch: FilterFetch {
   |                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

The blanket impl "solves" this problem though as it provides it for you. This causes the compiler to assume you missed the constraint that would allow the blanket impl to work.

**Comment by cart** [comment]
Thanks for solving that mystery. It was really bothering me :smile:

**Comment by cart** [comment]
Given that chalk integration isn't coming any time soon (afaik), I think we should probably just remove this comment and move on with our lives :)

**Comment by CGMossa** [comment]
Sure, fine by me.

**Pull request closed without merging**